File: backend/app/models/service_order.py
from datetime import datetime
from decimal import Decimal
from typing import Optional

from sqlalchemy import DateTime, ForeignKey, Numeric, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class ServiceOrder(Base):
    __tablename__ = "service_orders"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    client_id: Mapped[int] = mapped_column(
        ForeignKey("clients.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    equipment_id: Mapped[Optional[int]] = mapped_column(
        ForeignKey("equipments.id", ondelete="SET NULL"),
        index=True,
    )
    ticket_id: Mapped[Optional[int]] = mapped_column(
        ForeignKey("tickets.id", ondelete="SET NULL"),
        index=True,
    )
    assigned_user_id: Mapped[Optional[int]] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"),
        index=True,
    )
    opened_by_user_id: Mapped[Optional[int]] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"),
        index=True,
    )
    sold_by_user_id: Mapped[Optional[int]] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"),
        index=True,
    )
    number: Mapped[Optional[str]] = mapped_column(String(40), unique=True)
    title: Mapped[str] = mapped_column(String(180), nullable=False)
    status: Mapped[str] = mapped_column(String(30), nullable=False, default="aberta")
    priority: Mapped[str] = mapped_column(String(20), nullable=False, default="media")
    service_type: Mapped[Optional[str]] = mapped_column(String(80))
    received_equipment: Mapped[Optional[str]] = mapped_column(String(180))
    waiting_reason: Mapped[Optional[str]] = mapped_column(String(220))
    request_description: Mapped[str] = mapped_column(Text, nullable=False)
    technical_diagnosis: Mapped[Optional[str]] = mapped_column(Text)
    service_performed: Mapped[Optional[str]] = mapped_column(Text)
    internal_notes: Mapped[Optional[str]] = mapped_column(Text)
    labor_amount: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False, default=0)
    items_amount: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False, default=0)
    discount_amount: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False, default=0)
    total_amount: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False, default=0)
    opened_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )
    scheduled_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True))
    closed_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True))
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )

    client = relationship("Client", back_populates="service_orders")
    equipment = relationship("Equipment", back_populates="service_orders")
    ticket = relationship("Ticket", back_populates="service_orders")
    assigned_user = relationship(
        "User",
        back_populates="service_orders",
        foreign_keys=[assigned_user_id],
    )
    opened_by_user = relationship("User", foreign_keys=[opened_by_user_id])
    sold_by_user = relationship("User", foreign_keys=[sold_by_user_id])
    items = relationship(
        "ServiceOrderItem",
        back_populates="service_order",
        cascade="all, delete-orphan",
    )
    events = relationship(
        "ServiceOrderEvent",
        back_populates="service_order",
        cascade="all, delete-orphan",
        order_by="ServiceOrderEvent.created_at.desc()",
    )

    @property
    def opened_by_user_name(self) -> str | None:
        return self.opened_by_user.name if self.opened_by_user else None

    @property
    def opened_by_user_code(self) -> str | None:
        if self.opened_by_user is None:
            return None
        return self.opened_by_user.seller_code or self.opened_by_user.technician_code

    @property
    def assigned_user_name(self) -> str | None:
        return self.assigned_user.name if self.assigned_user else None

    @property
    def assigned_user_code(self) -> str | None:
        if self.assigned_user is None:
            return None
        return self.assigned_user.technician_code or self.assigned_user.seller_code

    @property
    def sold_by_user_name(self) -> str | None:
        return self.sold_by_user.name if self.sold_by_user else None

    @property
    def sold_by_user_code(self) -> str | None:
        if self.sold_by_user is None:
            return None
        return self.sold_by_user.seller_code or self.sold_by_user.technician_code


class ServiceOrderEvent(Base):
    __tablename__ = "service_order_events"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    service_order_id: Mapped[int] = mapped_column(
        ForeignKey("service_orders.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    user_id: Mapped[Optional[int]] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"),
        index=True,
    )
    event_type: Mapped[str] = mapped_column(String(40), nullable=False)
    status_from: Mapped[Optional[str]] = mapped_column(String(30))
    status_to: Mapped[Optional[str]] = mapped_column(String(30))
    assigned_user_id: Mapped[Optional[int]] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"),
        index=True,
    )
    notes: Mapped[Optional[str]] = mapped_column(Text)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )

    service_order = relationship("ServiceOrder", back_populates="events")
    user = relationship("User", foreign_keys=[user_id])
    assigned_user = relationship("User", foreign_keys=[assigned_user_id])

    @property
    def user_name(self) -> str | None:
        return self.user.name if self.user else None

    @property
    def assigned_user_name(self) -> str | None:
        return self.assigned_user.name if self.assigned_user else None

    @property
    def assigned_user_code(self) -> str | None:
        if self.assigned_user is None:
            return None
        return self.assigned_user.technician_code or self.assigned_user.seller_code


class ServiceOrderItem(Base):
    __tablename__ = "service_order_items"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    service_order_id: Mapped[int] = mapped_column(
        ForeignKey("service_orders.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    product_id: Mapped[Optional[int]] = mapped_column(
        ForeignKey("products.id", ondelete="SET NULL"),
        index=True,
    )
    description: Mapped[str] = mapped_column(String(180), nullable=False)
    quantity: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False, default=1)
    unit_price: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False, default=0)
    total_price: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False, default=0)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )

    service_order = relationship("ServiceOrder", back_populates="items")
    product = relationship("Product", back_populates="service_order_items")

File: admin_app/admin_flutter/lib/models/service_order.dart
class ServiceOrder {
  const ServiceOrder({
    required this.id,
    required this.clientId,
    required this.title,
    required this.status,
    required this.priority,
    required this.requestDescription,
    required this.laborAmount,
    required this.itemsAmount,
    required this.discountAmount,
    required this.totalAmount,
    required this.openedAt,
    required this.createdAt,
    required this.items,
    required this.events,
    this.equipmentId,
    this.ticketId,
    this.assignedUserId,
    this.openedByUserId,
    this.soldByUserId,
    this.number,
    this.serviceType,
    this.receivedEquipment,
    this.waitingReason,
    this.technicalDiagnosis,
    this.servicePerformed,
    this.internalNotes,
    this.scheduledAt,
    this.closedAt,
  });

  final int id;
  final int clientId;
  final int? equipmentId;
  final int? ticketId;
  final int? assignedUserId;
  final int? openedByUserId;
  final int? soldByUserId;
  final String? number;
  final String title;
  final String status;
  final String priority;
  final String? serviceType;
  final String? receivedEquipment;
  final String? waitingReason;
  final String requestDescription;
  final String? technicalDiagnosis;
  final String? servicePerformed;
  final String? internalNotes;
  final double laborAmount;
  final double itemsAmount;
  final double discountAmount;
  final double totalAmount;
  final DateTime openedAt;
  final DateTime? scheduledAt;
  final DateTime? closedAt;
  final DateTime createdAt;
  final List<ServiceOrderItem> items;
  final List<ServiceOrderEvent> events;

  factory ServiceOrder.fromJson(Map<String, dynamic> json) {
    return ServiceOrder(
      id: json['id'] as int,
      clientId: json['client_id'] as int,
      equipmentId: json['equipment_id'] as int?,
      ticketId: json['ticket_id'] as int?,
      assignedUserId: json['assigned_user_id'] as int?,
      openedByUserId: json['opened_by_user_id'] as int?,
      soldByUserId: json['sold_by_user_id'] as int?,
      number: json['number'] as String?,
      title: json['title'] as String,
      status: json['status'] as String,
      priority: json['priority'] as String,
      serviceType: json['service_type'] as String?,
      receivedEquipment: json['received_equipment'] as String?,
      waitingReason: json['waiting_reason'] as String?,
      requestDescription: json['request_description'] as String,
      technicalDiagnosis: json['technical_diagnosis'] as String?,
      servicePerformed: json['service_performed'] as String?,
      internalNotes: json['internal_notes'] as String?,
      laborAmount: _toDouble(json['labor_amount']),
      itemsAmount: _toDouble(json['items_amount']),
      discountAmount: _toDouble(json['discount_amount']),
      totalAmount: _toDouble(json['total_amount']),
      openedAt: DateTime.parse(json['opened_at'] as String),
      scheduledAt: json['scheduled_at'] == null
          ? null
          : DateTime.parse(json['scheduled_at'] as String),
      closedAt: json['closed_at'] == null
          ? null
          : DateTime.parse(json['closed_at'] as String),
      createdAt: DateTime.parse(json['created_at'] as String),
      items: (json['items'] as List<dynamic>? ?? [])
          .map(
            (item) => ServiceOrderItem.fromJson(item as Map<String, dynamic>),
          )
          .toList(),
      events: (json['events'] as List<dynamic>? ?? [])
          .map(
            (item) => ServiceOrderEvent.fromJson(item as Map<String, dynamic>),
          )
          .toList(),
    );
  }

  static double _toDouble(Object? value) {
    if (value == null) return 0;
    if (value is num) return value.toDouble();
    return double.tryParse(value.toString()) ?? 0;
  }
}

class ServiceOrderEvent {
  const ServiceOrderEvent({
    required this.id,
    required this.serviceOrderId,
    required this.eventType,
    required this.createdAt,
    this.userId,
    this.userName,
    this.statusFrom,
    this.statusTo,
    this.assignedUserId,
    this.assignedUserName,
    this.assignedUserCode,
    this.notes,
  });

  final int id;
  final int serviceOrderId;
  final int? userId;
  final String? userName;
  final String eventType;
  final String? statusFrom;
  final String? statusTo;
  final int? assignedUserId;
  final String? assignedUserName;
  final String? assignedUserCode;
  final String? notes;
  final DateTime createdAt;

  factory ServiceOrderEvent.fromJson(Map<String, dynamic> json) {
    return ServiceOrderEvent(
      id: json['id'] as int,
      serviceOrderId: json['service_order_id'] as int,
      userId: json['user_id'] as int?,
      userName: json['user_name'] as String?,
      eventType: json['event_type'] as String,
      statusFrom: json['status_from'] as String?,
      statusTo: json['status_to'] as String?,
      assignedUserId: json['assigned_user_id'] as int?,
      assignedUserName: json['assigned_user_name'] as String?,
      assignedUserCode: json['assigned_user_code'] as String?,
      notes: json['notes'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}

class ServiceOrderItem {
  const ServiceOrderItem({
    required this.id,
    required this.serviceOrderId,
    required this.description,
    required this.quantity,
    required this.unitPrice,
    required this.totalPrice,
    required this.createdAt,
    this.productId,
  });

  final int id;
  final int serviceOrderId;
  final int? productId;
  final String description;
  final double quantity;
  final double unitPrice;
  final double totalPrice;
  final DateTime createdAt;

  factory ServiceOrderItem.fromJson(Map<String, dynamic> json) {
    return ServiceOrderItem(
      id: json['id'] as int,
      serviceOrderId: json['service_order_id'] as int,
      productId: json['product_id'] as int?,
      description: json['description'] as String,
      quantity: ServiceOrder._toDouble(json['quantity']),
      unitPrice: ServiceOrder._toDouble(json['unit_price']),
      totalPrice: ServiceOrder._toDouble(json['total_price']),
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}

class ServiceOrderPayload {
  const ServiceOrderPayload({
    required this.clientId,
    required this.title,
    required this.status,
    required this.priority,
    required this.requestDescription,
    required this.laborAmount,
    required this.discountAmount,
    this.equipmentId,
    this.number,
    this.serviceType,
    this.receivedEquipment,
    this.waitingReason,
    this.technicalDiagnosis,
    this.servicePerformed,
    this.internalNotes,
    this.scheduledAt,
  });

  final int clientId;
  final int? equipmentId;
  final String? number;
  final String title;
  final String status;
  final String priority;
  final String? serviceType;
  final String? receivedEquipment;
  final String? waitingReason;
  final String requestDescription;
  final String? technicalDiagnosis;
  final String? servicePerformed;
  final String? internalNotes;
  final double laborAmount;
  final double discountAmount;
  final DateTime? scheduledAt;

  Map<String, dynamic> toJson() {
    return {
      'client_id': clientId,
      'equipment_id': equipmentId,
      'number': _emptyToNull(number),
      'title': title.trim(),
      'status': status,
      'priority': priority,
      'service_type': _emptyToNull(serviceType),
      'received_equipment': _emptyToNull(receivedEquipment),
      'waiting_reason': _emptyToNull(waitingReason),
      'request_description': requestDescription.trim(),
      'technical_diagnosis': _emptyToNull(technicalDiagnosis),
      'service_performed': _emptyToNull(servicePerformed),
      'internal_notes': _emptyToNull(internalNotes),
      'labor_amount': laborAmount,
      'discount_amount': discountAmount,
      'scheduled_at': scheduledAt?.toUtc().toIso8601String(),
    };
  }
}

class ServiceOrderItemPayload {
  const ServiceOrderItemPayload({
    required this.description,
    required this.quantity,
    required this.unitPrice,
    this.productId,
  });

  final int? productId;
  final String description;
  final double quantity;
  final double unitPrice;

  Map<String, dynamic> toJson() {
    return {
      'product_id': productId,
      'description': description.trim(),
      'quantity': quantity,
      'unit_price': unitPrice,
    };
  }
}

String? _emptyToNull(String? value) {
  final trimmed = value?.trim();
  if (trimmed == null || trimmed.isEmpty) return null;
  return trimmed;
}
